packages/localize-tools/src/tests/analysis.unit.test.ts

```diff
@@ -347,6 +347,22 @@ test('error: different message contents', () => {
   );
 });
 
+test('same message contents with different expression is not error', () => {
+  const src = `
+    import {msg, str} from '@lit/localize';
+    const name = "friend";
+    const otherName = "pal";
+    msg(str\`Hello \${name}\`);
+    msg(str\`Hello \${otherName}\`);
+  `;
+  checkAnalysis(src, [
+    {
+      name: 'saed7d3734ce7f09d',
+      contents: ['Hello ', {untranslatable: '${name}'}],
```

Comment:
Hmm... Seeing this example has actually made me realize there is a bit of a problem with this change.

Take this example:

```ts
msg(`Hello ${friend}`)
msg(`Hello ${pal}`)
```

Both msg calls will map to the same message id, so there will only be one XLIFF entry for both calls, which will look like this:

```xml
<trans-unit id="saed7d3734ce7f09d">
  <source>Hello <x equiv-text="${friend}"/></source>
</trans-unit>
```

Note that we put the actual source code of the expression inside the `<x>` tag. This is because we want to provide some information to the translator about what that `<x>` placeholder means, so that they can decide where it should go in the translated version (translators can move, but not change `<x>` tags). 

But, we had to pick one of the expressions from the two `msg` calls that correspond to this message id. I expect that currently we'd whichever one we encounter first, which is determined by the order that we traversed the module graph.

This is already a bit of a problem, because it means changing around the module graph could cause surprising churn in the generated XLIFF files, which would be noisy and could trigger unnecessary re-translation. Ideally this would be stable and predictable.

We could prefer shorter expressions? Lexicographic?

---

The larger problem, though, is when we go to produce our output in transform mode.

Let's say we have two `msg` calls with 2 expressions, that get assigned the same id:

```ts
msg(`foo ${a} bar {b}`)
msg(`foo ${c} bar {d}`)
```

So as before, the first `msg` call we encountered with that id wins as the "canonical" one, and we generate XLIFF like this:

```xml
<source>foo <x equiv-text="${a}"/> bar <x equiv-text="${b}"/></target>
```

Now let's say we get a translation back where the order of the `<x>` tags was reversed (which could happen for reasons like a difference in subject/verb/object ordering between two languages):

```xml
<target>xxx <x equiv-text="${b}"/> yyy <x equiv-text="${a}"/></target>
```

In transform mode, we would _want_ to generate this:

```ts
msg(html`xxx <b>${b}</b> yyy {a}`)
msg(html`xxx <b>${d}</b> yyy {c}`)
```

But... we'll hit a problem. When generating translated `msg` calls, we currently rely on the fact that the expressions in the source file text will match the ones in the `equiv-text` XLIFF attributes. This is because we need to be able to find the AST node that corresponds to each expression, and insert it into the placeholders in the newly generated translated template.

This logic happens here: https://github.com/lit/lit/blob/5ac025bf9610adb7069ef8d88ed6bd96ff730f2f/packages/localize-tools/src/modes/transform.ts#L374

The problem is that in the second `msg` call, the source file text won't match the `equiv-text` attributes. We're trying to find the positions of "d" and "c", but we only see "a" and "b". We didn't previously have this problem because we enforced that the source text for all expressions were identical whenever two `msg` calls had the same id.

---

So, I think we'd need to do something smarter here to make this work, where instead of mapping `source expression -> xliff expression`, we instead do `source expression -> canonical source expression -> xliff expression`. We should be able to do the new `source -> canonical` mapping by using the _offset_ of the expressions.

So in this case we would figure out that "c" in the second `msg` call corresponds to "a" in the first `msg` call because it is the first expression. Same idea for "d" corresponding to "b". Now we have the _canonical_ expression values, and we will be able to match the `equiv-text` attributes correctly.

Reply:
I see. So if I understood this correctly, we need to decide on some rule to determine what the canonical source expressions are going to be so the generated xliff is stable, but also we need to account for the ordering of the expression and be able to map that to a different set of source expressions from a message with the same id for transform mode. (For runtime mode this isn't an issue because we're already replacing expressions with indexes?)

Reply:
> I see. So if I understood this correctly, we need to decide on some rule to determine what the canonical source expressions are going to be so the generated xliff is stable, but also we need to account for the ordering of the expression and be able to map that to a different set of source expressions from a message with the same id for transform mode. (For runtime mode this isn't an issue because we're already replacing expressions with indexes?)

Yep, I think that's exactly right!

It would be nice to include a test that verifies that runtime mode also works when a translation changes expression order, but I think you're correct that it will be unaffected by this issue.